### PR TITLE
kernel: support memory mapped stacks

### PR DESCRIPTION
**Pull request by dcpleung**
This adds support to memory map thread stacks during thread creation, and subsequent (delayed) un-mapping when the thread is terminated. With memory mapped stacks, guard pages can be created enclosing the stacks to catch stack overflow and underflow. This requires MMU.

(See individual commits for details.)

Relates: #28899

### CODE REVIEW

**Review by cfriedt — APPROVED** [review]
Well done! It's not every day a 2xxxx series issue gets some attention 🪁

File: kernel/thread.c [hunk 1]
```diff
@@ -939,3 +985,113 @@ int k_thread_runtime_stats_all_get(k_thread_runtime_stats_t *stats)
...
+	thread_cleanup_stack_addr = thread->stack_info.mapped.addr;
+	thread_cleanup_stack_sz = thread->stack_info.mapped.sz;
+
+	/* The stack is now considered un-usable. This should prevent any functions
+	 * from looking directly into the mapped stack if they are made to be aware
+	 * of memory mapped stacks, e.g., z_stack_space_get().
+	 */
+	thread->stack_info.mapped.addr = NULL;
+	thread->stack_info.mapped.sz = 0;
+#endif /* CONFIG_THREAD_STACK_MEM_MAPPED */
```
Comment (ceolin) [inline kernel/thread.c:1030]: How this works with smp ? What happens if two threads running in parallel do `z_thread_abort(self)` ?

**Comment by ceolin** [comment]
yay ! overflow detection !

**Review by jhedberg — APPROVED** [review]

File: kernel/thread.c (hunk 1 above)
Comment (dcpleung) [reply kernel/thread.c:1030]: The spinlock in `k_thread_abort_cleanup()` will serialize them. Thread 1 will setup the data for deferred unmapping. Thread 2 will see that and will do the unmapping for thread 1, then setup its own deferred unmapping data.

Comment (ceolin) [reply kernel/thread.c:1030]: aaah ok, there is a cleanup first for any pending thread.

**Review by ceolin — APPROVED** [review]

**Comment by andyross** [comment]
I'm gonna be that guy again.  This is a huge patch and a ton of complexity... do we really want to do this this way?  The feature has value, but is this the right API?

I mean, look at Linux.   Linux has all this stuff too.  And it doesn't have *any* support in the kernel to do it.  You create a new thread stack with mmap() and just point clone() to it, and the kernel uses it right where you told it to.

Can't we do this with a library layer where you say "give me a mmap'd thread stack", and you get one, and then pass that to k_thread_create()?  Why bother making a complicated thread creation API even more complicated?

**Comment by cfriedt** [comment]
This could also be used for some other interesting things aside from thread stacks 😏

File: kernel/include/kernel_internal.h [hunk 2]
```diff
@@ -297,6 +297,35 @@ int z_kernel_stats_raw(struct k_obj_core *obj_core, void *stats);
...
+ * where these steps require that the thread is no longer running.
+ * If the target thread is not the current running thread, the cleanup
+ * steps will be performed immediately. However, if the target thread is
+ * the current running thread (e.g. k_thread_abort(_current)), it defers
+ * the cleanup steps to later when the work will be finished in another
+ * context.
+ *
+ * @param thread Pointer to thread to be cleaned up.
+ */
+void k_thread_abort_cleanup(struct k_thread *thread);
```
Comment (andyross) [inline kernel/include/kernel_internal.h:314]: Good example here: if your stack creation is a library level thing, you can just provide a top level wrapper or whatever and clean it up on exit.  No need to muck with thread lifecycle stuff in the kernel

File: kernel/thread.c [hunk 3]
```diff
@@ -405,11 +405,47 @@ static char *setup_thread_stack(struct k_thread *new_thread,
...
+	 * calculate the initial stack pointer at the high end of the stack
+	 * object. The stack pointer may be reduced later in this function
+	 * by TLS or random offset.
+	 *
+	 * K_MEM_MAP_UNINIT is used to mimic the behavior of non-mapped
+	 * stack. If CONFIG_INIT_STACKS is enabled, the stack will be
+	 * cleared below.
+	 */
+	void *stack_mapped = k_mem_phys_map((uintptr_t)stack, stack_obj_size,
+					    K_MEM_PERM_RW | K_MEM_CACHE_WB | K_MEM_MAP_UNINIT);
```
Comment (andyross) [inline kernel/thread.c:419]: Also just to be pedantic, this isn't an optional thing in this scheme:  every stack in every thread gets remapped automatically whether it wants it or not.  But control over stack memory is something a lot of apps want.  Consider a runtime for a Go-like language with lots of I/O threads flying around using tiny bounded stacks, etc...

Obviously most apps will just want defaults, but at the same time most apps would be just as happy with a recommendation like "use pthreads to get automatic MMU stack bounds, k_thread_create() is the low level API".

**Review by andyross — COMMENTED** [review]
> This could also be used for some other interesting things aside from thread stacks 😏

In fact pthreads would be the obvious choice for the "wrapper" API that does the appropriate stack allocation/mapping magic on behalf of the app, because that's where users expect that abstraction to live as that's where they find it on Linux.

**Comment by dcpleung** [comment]
This is done in this way to be as transparent to the API user as possible. As long as the kconfig is enabled, the feature is enabled without any code changes on the application.

With your idea of having a library call to produce memory mapped stacks (whether coming from malloc or statically declared stacks), the user of `k_thread_create()` has to be explicitly call this library beforehand which requires changes to any code trying to use this feature. Even a thread option to enable this feature on a per thread basis would result in similar code to the current form since it still needs to do things behind the scene.

If you can come up with another solution without requiring application code changes, please feel free to submit it. I am very interested in seeing such solution.

**Comment by andyross** [comment]
But... why do we not want application code changes, though?   New features get new APIs, that's not so weird to me?  And I repeat, it's not as simple as "just turn it on", some apps **won't** want this, or will want it for only certain threads.  k_thread_create() is an extremely complicated API already.

The philosophy here is basically that toolkits are better than frameworks.  There's no Right Answer for how to lay out memory spaces, it's just a complicated area.  Every runtime on Linux tends to have its own ideas about how to do this, for example (Python threads don't get initialized like bare pthreads, which are different from Rust, etc...) and it won't be any different for us.  If we land this API like it is we're stuck with it forever, and I really don't think this is likely to evolve cleanly.

**Comment by cfriedt** [comment]
Just a suggestion, couldn't mapping be a flag that is added to k_thread_stack_alloc()?

I assume it should be possible to detect if a stack is mapped in k_thread_stack_free()

**Comment by dcpleung** [comment]
If an app doesn't want it, they don't have to turn on the kconfig. I can add a thread creation flag to use stack as-is. The fundamental question is, whether we want to be safe by default, or only opt into safety if needed (or rather, if someone remembers to be safe). I would argue that in most cases, safe by default should be the choice. And that being unsafe needs to be a conscious choice (or rather, MUST be a conscious choice).

Here, having guards around stack serves is important to prevent corrupting data when running in kernel mode, whether with kernel mode only threads or serving syscalls with privileged stack. Overflowing thread stack in user mode will touch the privileged stack in current arch MMU code and it will generate page faults. So that is covered. However, with kernel mode only threads and serving syscalls, they are running in kernel mode. And given that most stacks are declared statically in the `.noinit` section, any overflowing of stack will corrupt other data. This is because under kernel mode, everything is accessible.

**Comment by dcpleung** [comment]
> Just a suggestion, couldn't mapping be a flag that is added to k_thread_stack_alloc()?
> 
> I assume it should be possible to detect if a stack is mapped in k_thread_stack_free()

This only works if stacks are allocated at runtime. For statically declared stacks (which majority of them are), they would still need to be pre-processed first if mapping is not handled automatically.

**Pull request merged**